From 5be2a945ca16777f6fc5f38d926f104a07b8c338 Mon Sep 17 00:00:00 2001
From: Richard Yu <richard.yu@mitrastar.com.tw>
Date: Tue, 20 May 2025 11:18:46 +0800
Subject: [PATCH] controller and agent: enhance multiple controllers detection
 via AP-Autoconfiguration Search

This commit updates the behavior of Multi-AP devices that function
as both a Multi-AP Agent and a Multi-AP Controller. These devices
now include "Multi-AP Controller" in the SupportedService TLV of the
AP-Autoconfiguration Search message, in addition to the previously
included "Multi-AP Agent".

The AP-Autoconfiguration Search message, being a relayed multicast,
reaches all other Multi-AP devices on the network. When another
Multi-AP Controller receives this message, it can detect the presence
of multiple controllers and logs the following message:

LOG(ERROR) << "[Multiple Controllers Detected] Received AP_AUTOCONFIGURATION_SEARCH_MESSAGE from another Controller: " << al_mac;

This enhancement introduces a more efficient mechanism for detecting
multiple controllers by leveraging the AP-Autoconfiguration Search
message, supplementing the existing approach that relies solely on
unicast 1905 messages, as employed in earlier development commits
associated with PPM-2373.

It is important to note that the Wi-Fi EasyMesh Specification does
not mandate the inclusion of "Multi-AP Controller" in the
SupportedService TLV of the AP-Autoconfiguration Search message.
Therefore, this enhancement is considered a custom extension
aimed at improving network diagnostics.

Despite being a custom extension, this change maintains
interoperability with other prplMesh devices that do not include this
enhancement. Devices that do not expect the additional
"Multi-AP Controller" indication in the SupportedService TLV will
simply ignore it, ensuring seamless operation within mixed-version
prplMesh environments.

https://prplfoundationcloud.atlassian.net/browse/PPM-3314

Signed-off-by: Richard Yu <richard.yu@mitrastar.com.tw>
---
 .../slave/tasks/ap_autoconfiguration_task.cpp       | 12 ++++++++++--
 controller/src/beerocks/master/controller.cpp       | 13 ++++++++++++-
 2 files changed, 22 insertions(+), 3 deletions(-)
 mode change 100644 => 100755 controller/src/beerocks/master/controller.cpp

diff --git a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
index 03c87a4b31..323c34a55b 100644
--- a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
+++ b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
@@ -507,6 +507,9 @@ bool ApAutoConfigurationTask::send_ap_autoconfiguration_search_message(
     auto db = AgentDB::get();
 
     auto numberOfSupportedService = 2;
+    if (db->device_conf.local_controller) {
+        numberOfSupportedService += 2;
+    }
     if (db->device_conf.certification_mode) {
         // The R4 QCA WFA controller won't answer Autoconfig Searches that specify EM_AP_AGENT as a supported service. See PPM-3287.
         numberOfSupportedService = 1;
@@ -589,10 +592,15 @@ bool ApAutoConfigurationTask::send_ap_autoconfiguration_search_message(
             if (serviceID == 0) {
                 std::get<1>(supportedServiceTuple) =
                     wfa_map::tlvSupportedService::eSupportedService::MULTI_AP_AGENT;
-            }
-            if (serviceID == 1) {
+            } else if (serviceID == 1) {
                 std::get<1>(supportedServiceTuple) =
                     wfa_map::tlvSupportedService::eSupportedService::EM_AP_AGENT;
+            } else if (serviceID == 2) {
+                std::get<1>(supportedServiceTuple) =
+                    wfa_map::tlvSupportedService::eSupportedService::MULTI_AP_CONTROLLER;
+            } else if (serviceID == 3) {
+                std::get<1>(supportedServiceTuple) =
+                    wfa_map::tlvSupportedService::eSupportedService::EM_AP_CONTROLLER;
             }
         }
 
diff --git a/controller/src/beerocks/master/controller.cpp b/controller/src/beerocks/master/controller.cpp
old mode 100644
new mode 100755
index 1afc2b897b..3ec00d79a5
--- a/controller/src/beerocks/master/controller.cpp
+++ b/controller/src/beerocks/master/controller.cpp
@@ -677,7 +677,8 @@ bool Controller::handle_cmdu_1905_autoconfiguration_search(const sMacAddr &src_m
     auto &auto_config_freq_band = tlvAutoconfigFreqBand->value();
     LOG(DEBUG) << "band=" << int(auto_config_freq_band);
 
-    bool supported_agent_service = false;
+    bool supported_controller_service = false;
+    bool supported_agent_service      = false;
     for (int i = 0; i < tlvSupportedServiceIn->supported_service_list_length(); i++) {
         auto supportedServiceTuple = tlvSupportedServiceIn->supported_service_list(i);
         if (!std::get<0>(supportedServiceTuple)) {
@@ -687,11 +688,21 @@ bool Controller::handle_cmdu_1905_autoconfiguration_search(const sMacAddr &src_m
         auto supportedService = std::get<1>(supportedServiceTuple);
         if (supportedService == wfa_map::tlvSupportedService::eSupportedService::MULTI_AP_AGENT) {
             supported_agent_service = true;
+        } else if (supportedService ==
+                       wfa_map::tlvSupportedService::eSupportedService::MULTI_AP_CONTROLLER ||
+                   supportedService ==
+                       wfa_map::tlvSupportedService::eSupportedService::EM_AP_CONTROLLER) {
+            supported_controller_service = true;
         } else {
             LOG(INFO) << "Not supported service, received value:" << std::hex
                       << int(supportedService);
         }
     }
+    if (supported_controller_service && al_mac != database.get_local_bridge_mac()) {
+        LOG(ERROR) << "[Multiple Controllers Detected] Received "
+                      "AP_AUTOCONFIGURATION_SEARCH_MESSAGE from another Controller: "
+                   << al_mac;
+    }
     if (!supported_agent_service) {
         LOG(WARNING) << "MULTI_AP_AGENT is not supported as service";
         return false;
-- 
GitLab

